mm2src/coins/eth/eth_swap_v2.rs

```diff
@@ -96,14 +129,327 @@ impl EthCoin {
+            taker_swap_v2_contract,
+            taker_status,
+            TakerPaymentStateV2::PaymentSent as u8,
+        )?;
+
+        let validation_args = {
+            let dex_fee = wei_from_big_decimal(&args.dex_fee.fee_amount().into(), self.decimals)?;
+            let payment_amount = wei_from_big_decimal(&(args.trading_amount + args.premium_amount), self.decimals)?;
+            TakerValidationArgs {
+                swap_id,
+                amount: payment_amount,
+                dex_fee,
+                receiver: self.my_addr().await,
+                taker_secret_hash: args.taker_secret_hash,
+                maker_secret_hash: args.maker_secret_hash,
+                funding_time_lock,
+                payment_time_lock,
+            }
+        };
+        match self.coin_type {
+            EthCoinType::Eth => {
+                let function = TAKER_SWAP_V2.function(ETH_TAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_eth_taker_payment_data(&decoded, &validation_args, function, tx_from_rpc.value)?;
+            },
+            EthCoinType::Erc20 { token_addr, .. } => {
+                let function = TAKER_SWAP_V2.function(ERC20_TAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_erc20_taker_payment_data(&decoded, &validation_args, function, token_addr)?;
+            },
+            EthCoinType::Nft { .. } => unreachable!(),
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn taker_payment_approve(
+        &self,
+        args: &GenTakerFundingSpendArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let (taker_swap_v2_contract, send_func, token_address) = self
+            .taker_swap_v2_details(ETH_TAKER_PAYMENT, ERC20_TAKER_PAYMENT)
+            .await?;
+        // Note: `PaymentSent` state was checked in `validate_taker_funding`. So we don't have to validate it in prepare function
+        let data = {
+            let decoded = try_tx_s!(decode_contract_call(send_func, args.funding_tx.unsigned().data()));
+            try_tx_s!(
+                self.prepare_taker_payment_approve_data(args, decoded, token_address)
+                    .await
+            )
+        };
+        // TODO need new consts and params for v2 calls, here should be one `gas_limit.taker_approve` param without match
+        let gas_limit = match self.coin_type {
+            EthCoinType::Eth => U256::from(self.gas_limit.eth_payment),
+            EthCoinType::Erc20 { .. } => U256::from(self.gas_limit.eth_payment),
+            EthCoinType::Nft { .. } => unreachable!(), // EthCoinType::Nft case was already checked in taker_swap_v2_details
+        };
+        let approve_tx = self
+            .sign_and_send_transaction(0.into(), Action::Call(taker_swap_v2_contract), data, gas_limit)
+            .compat()
+            .await?;
+        Ok(approve_tx)
+    }
```

Comment:
we should check `coin_type` at the beginning of this function too
```rust
        if let EthCoinType::Nft { .. } = self.coin_type {
             return MmError::err(ValidateSwapV2TxError::Internal(
                 "EthCoinType must be ETH or ERC20".to_string(),
             ));
         }
```

Reply:
done https://github.com/KomodoPlatform/komodo-defi-framework/pull/2169/commits/9f3164995bd463227c570ad9bbd890e3103142ec

ps: but i did use unreachable in private `taker_swap_v2_details` func for Nft case, as it "doesnt know" where it is called and what is already checked.